src/output-writer/csv-writer.ts

```diff
@@ -0,0 +1,117 @@
+import * as fs from 'fs/promises';
+import type { OutputWriter } from './index.js';
+
+const enum CSVColumnType {
+  CategoryScore,
+  AuditScore,
+  AuditValue,
+}
+
+type CSVColumnField =
+  | {
+      type: CSVColumnType.AuditScore;
+      audit: string;
+    }
+  | {
+      type: CSVColumnType.AuditValue;
+      audit: string;
+    }
+  | {
+      type: CSVColumnType.CategoryScore;
+      category: string;
+    };
+
+interface CSVColumn {
+  name: string;
+  category: string;
+  field: CSVColumnField;
+}
+
+const makeCSVRow = (cells: string[]) => `${cells.join(',')}\n`;
+
+export const createCSVOutputWriter = async (
+  filePath: string
+): Promise<OutputWriter> => {
+  const outputFile = await fs.open(filePath, 'w');
+  let hasWrittenHeader = false;
+  // Used to make sure that the addEntry calls happen one at a time
+  // so they always write to the file in a deterministic order
+  // (specifically important to make sure the header is first)
+  // This promise chains all the promises from addEntry calls,
+  // making sure each call finishes before the next one is processed
+  let mutexPromise: Promise<unknown> = Promise.resolve();
+  const columns: CSVColumn[] = [];
+  return {
+    async addEntry(report) {
+      mutexPromise = mutexPromise.then(async () => {
+        if (!hasWrittenHeader) {
+          hasWrittenHeader = true;
+          for (const category of Object.values(report.categories)) {
+            columns.push({
+              name: `${category.title} (overall score)`,
```

Comment:
Not a change request for this PR, but seeing "overall score" hardcoded in English made me wonder if the tool could support different language outputs in the future. 🙂

Reply:
fun to read these in Spanish: https://github.com/microsoft/TypeScript/blob/main/src/loc/lcl/esn/diagnosticMessages/diagnosticMessages.generated.json.lcl#L463